docs/proposals/high_availabilty_azure.md

```diff
@@ -0,0 +1,236 @@
+
+Pros 
+
+1. We will have high availability as there will be multiple guard pods which can receive and respond to the checkaccess requests that are coming. Thus, a sudden spike of requests or a continuous flow of a high number of requests can be easily handled. 
+
+2. As the kube-apiserver cache will be disabled, the no of requests will increase so having more than one replica of guard will be useful to handle the request load. 
+
+Cons: 
+
+1. Eventual consistency of cache and depending on the approach we take based on the below sections, latency might increase too. 
+
+## Cache in Multiple Replica Guard Scenario 
+
+Currently in guard we use an in-memory cache (Big Cache) to store the results of the RBAC decision sent by the checkaccess call. This result is cached for 5 mins. Without a cache, guard would need to send a checkaccess request to ARM for each review request that comes from kube-apiserver. This would have resulted in increased network latency and throttling at the PAS layer. Also, if there were consecutive requests for the same resource and user from kube-apiserver, guard would have to make a checkaccess call for each of those requests. 
+
+The result is cached only for 5 mins as the RBAC of the resource for a particular user can change depending on the role assignment changes made. Hence, we do not want to store the result permanently. As PAS recommends that check access result is valid for 5 min and RP/client should do another call after 5 min. It returns ttl in check access response which is set to 5 min as of now for all responses. Hence, we are storing the cache result for 5 mins. In the multiple guard replica scenario, we would still need to have cache as the no of requests to guard will increase as the kube-apiserver cache will be disabled. 
+
+### Objectives 
+
+The cache needs to be consistent with all replicas as we want to have only one version of the result and the result should be invalidated in 5 mins. 
+
+The cache lifetime can be tied to the pod lifetime as we are not looking for permanent storage. 
+
+### Caching strategies 
+
+#### Sidecar Pattern using Hazelcast 
+
+Ideally, we want a low latency, fast in-memory distributed cache known as an in memory distributed cache(IMDG). But all the embedded distributed cache offerings are in JVM, hence we are going with a sidecar pattern with an embedded distributed cache. Sidecar topology brings the benefits of Embedded topology because Hazelcast scales together with the application and both containers run on the same machine. However, the application can be written in any programming language because it connects to Hazelcast members using the standard Hazelcast Client libraries which are available in Go. We investigated Redis as a sidecar as well, but it does support distributed cache across sidecars thus will not have a consistent cache across all the replicas. 
```

Comment:
> We investigated Redis as a sidecar as well, but it does support distributed cache across sidecars thus will not have a consistent cache across all the replicas. 

I suppose you want to say "We investigated Redis as a sidecar as well, but it does **NOT** support distributed cache across sidecars thus will not have a consistent cache across all the replicas."?